### PR TITLE
[FEATURE] Course age filtering  #42aeqt

### PR DESCRIPTION
**Pull request by jcoryalvernaz**
# Overview

Add functionality to support filtering listed programs in a given location by a specific age range.

## Details
* Add an Array.filter method to filter based on selected age filter (< 5, 5, 6, 7, 8, 9, or > 10)
* Dynamically populate the drop down list for only those age ranges that are available for the given list of courses.
* Apply filter to the list of course being displayed when drop down item is selected.
* Update the listings counters to reflect the correct number of courses displayed.
* Lock drop down list if only one age filter item is returned.
* Update the URL to show filters that are currently being applied.

## QA Instructions

* Visit the [programs page](https://5e488ce4dbc01a0007de9e1d--play-well-staging.netlify.com/programs/) on the Playwell staging site.
* Enter in a location.
* Selecting an age from the drop down should update the list of courses for that age range.
* Selecting an age from the drop down should update the label of the drop down.
* Selecting an age from the drop down should update the course counters below the drop down.
* Selecting an age from the drop down should update the url with the query parameters for `age_min` and `age_max`

[#42aeqt](https://app.clickup.com/t/42aeqt)

### CODE REVIEW

File: web/src/components/library/CourseListings/CourseListings.js [hunk 1]
```diff
@@ -365,19 +368,33 @@ class ListingsResults extends PureComponent {
...
-            courses: node.node.courses.filter((course, idxx) => {
-              // Define our query
-              const filter = d;
-
-              // Set up the conditions
-              return course.category_group_name.includes(filter);
-            }, this),
+            courses: node.node.courses
+              .filter(course => course.category_group_name.includes(filter))
+              .filter(course  => filteredCourseByAge(course)),
```
[the remark this answers is not in the document]
Comment (jcoryalvernaz) [reply web/src/components/library/CourseListings/CourseListings.js]: Nice, that is cleaner.

File: web/src/components/library/CourseListings/CourseListings.js [hunk 2]
```diff
@@ -365,19 +368,33 @@ class ListingsResults extends PureComponent {
         }
       }, this);
 
-    const filteredCourseDataByToggle = d =>
+    const filteredCourseByAge = course => {
```
[the remark this answers is not in the document]
Comment (jcoryalvernaz) [reply web/src/components/library/CourseListings/CourseListings.js]: Yeah this function kind of bugs me too. I'll get it formatted better.

File: web/src/components/library/CourseListings/ListingsFilters/ListingsFilters.js [hunk 3]
```diff
@@ -7,6 +7,7 @@
 // Core
 import React from 'react';
 import { Link } from 'gatsby';
+import { range, uniqWith, isEqual, flatten } from 'lodash'
```
[the remark this answers is not in the document]
Comment (jcoryalvernaz) [reply web/src/components/library/CourseListings/ListingsFilters/ListingsFilters.js]: That makes sense. I made this change. Left the `range` function as it is being called in the `getAllAges` function.

File: web/src/components/library/CourseListings/ListingsFilters/ListingsFilters.js [hunk 4]
```diff
@@ -21,40 +22,64 @@ import { ListingsFiltersStyle } from './styles.scss';
...
+    const sortedAges = mergedAgeList.sort((a,b) => (a-b));
+    return sortedAges
+  })
+
+  const createAgeFilterItems = courses => {
+    const ageList = getAllAges(courses)
+    const ageFilterItems = uniqWith(ageList
+      .map(age => {
+        if (age < 5) {
+          return { name: `Under 5`, value: { ageMin: 0, ageMax: 5} }
```
[the remark this answers is not in the document]
Comment (jcoryalvernaz) [reply web/src/components/library/CourseListings/ListingsFilters/ListingsFilters.js]: Agreed.

File: web/src/components/library/CourseListings/ListingsFilters/ListingsFilters.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (elskwid) [reply web/src/components/library/CourseListings/ListingsFilters/ListingsFilters.js]: @jcoryalvernaz I must have missed the `range` usage. Sorry for the confusion! I def see it now. 🙃

**Comment by elskwid** [comment]
Nice updates! Thanks.

**Comment by jrossm** [comment]
QA Failed on this one --

🆗 Selecting an age from the drop down updates list of courses for that age range
🔴 Selecting an age from the drop down doesn't update the drop down label
🆗 Selecting an age from the drop down updates course counters
🆗 Selecting an age from the drop down updates the url with the query parameters for `age_min` and `age_max`

🎁 Bonus bug I found 🎁 
the drop down disappears too quickly certain locations (Alabama in this case) & after multiple attempts to interact with it the page refreshes to a blank screen
video of the bug in action here: 
https://monosnap.com/file/BpyS9WGjMTEqnRhlDiz4subDy6pBsX

**Comment by jcoryalvernaz** [comment]
@jrossm I mistakenly put the drop down label update on this PR. This feature is actually in PR #11. I was not able to reproduce the bug you showed, but I will put it in as a task so I don't forget it. Thanks

**Comment by jrossm** [comment]
QA passed! ✨ @jcoryalvernaz I tried to reproduce that bug, and can only do so on safari

**Pull request merged**